iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/resource/memory/InsertNodeMemoryEstimator.java

```diff
@@ -0,0 +1,1111 @@
+    size += sizeOfTEndPoint(tDataNodeLocation.getSchemaRegionConsensusEndPoint());
+
+    return size;
+  }
+
+  private static long sizeOfTEndPoint(final TEndPoint tEndPoint) {
+    if (tEndPoint == null) {
+      return 0L;
+    }
+    // objectHeader + ip + port
+    long size = T_END_POINT_SIZE;
+
+    size += sizeOfString(tEndPoint.ip);
+    return size;
+  }
+
+  private static long sizeOfTSStatus(final TSStatus tSStatus) {
+    if (tSStatus == null) {
+      return 0L;
+    }
+    long size = TS_STATUS_SIZE;
+    // message
+    if (tSStatus.isSetMessage()) {
+      size += sizeOfString(tSStatus.message);
+    }
+    // ignore subStatus
+    // redirectNode
+    if (tSStatus.isSetRedirectNode()) {
+      size += sizeOfTEndPoint(tSStatus.redirectNode);
+    }
+    return size;
+  }
+
+  // =============================ProgressIndex==================================
+
+  private static long sizeOfProgressIndex(final ProgressIndex progressIndex) {
+    if (progressIndex == null) {
+      return 0L;
+    }
+    switch (progressIndex.getType()) {
+      case HYBRID_PROGRESS_INDEX:
+        return sizeOfHybridProgressIndex((HybridProgressIndex) progressIndex);
+      case IOT_PROGRESS_INDEX:
+        return sizeOfIoTProgressIndex((IoTProgressIndex) progressIndex);
+      case META_PROGRESS_INDEX:
+        return sizeOfMetaProgressIndex();
+      case STATE_PROGRESS_INDEX:
+        return sizeOfStateProgressIndex((StateProgressIndex) progressIndex);
+      case SIMPLE_PROGRESS_INDEX:
+        return sizeOfSimpleProgressIndex();
+      case MINIMUM_PROGRESS_INDEX:
+        return sizeOfMinimumProgressIndex();
+      case RECOVER_PROGRESS_INDEX:
+        return sizeOfRecoverProgressIndex((RecoverProgressIndex) progressIndex);
+      case TIME_WINDOW_STATE_PROGRESS_INDEX:
+        return sizeOfTimeWindowStateProgressIndex((TimeWindowStateProgressIndex) progressIndex);
+    }
+    return 0L;
+  }
+
+  private static long sizeOfHybridProgressIndex(final HybridProgressIndex progressIndex) {
+    // Memory alignment of basic types and reference types in structures
+    long size = HYBRID_PROGRESS_INDEX_SIZE;
+
+    // Memory calculation in reference type, cannot get exact value, roughly estimate
+    size += REENTRANT_READ_WRITE_LOCK_SIZE;
+    if (progressIndex.getType2Index() != null) {
+      size +=
+          NUM_BYTES_OBJECT_HEADER
+              + progressIndex.getType2Index().size() * (Short.BYTES + NUM_BYTES_OBJECT_REF);
+    }
+    return size;
+  }
+
+  private static long sizeOfIoTProgressIndex(IoTProgressIndex progressIndex) {
+    // Memory alignment of basic types and reference types in structures
+    long size = IO_T_PROGRESS_INDEX_SIZE;
+
+    // Memory calculation in reference type, cannot get exact value, roughly estimate
+    size += REENTRANT_READ_WRITE_LOCK_SIZE;
+
+    size +=
+        NUM_BYTES_OBJECT_HEADER
+            + progressIndex.getPeerId2SearchIndexSize() * (long) (Integer.BYTES + Long.BYTES);
```

Comment:
Maybe need to add a RamUsageEstimator.HASHTABLE_RAM_BYTES_PER_ENTRY? The size of Map.Entry may not be simply a "keySize + valueSize"....